src/modules/rlm_sql/rlm_sql.c

```diff
@@ -935,64 +935,88 @@ static int sql_groupcmp(void *instance, REQUEST *request, UNUSED VALUE_PAIR *req
 	return 1;
 }
 
-static rlm_rcode_t rlm_sql_process_groups(rlm_sql_t const *inst, REQUEST *request, rlm_sql_handle_t **handle,
-					  sql_fall_through_t *do_fall_through)
+static rlm_rcode_t rlm_sql_process_groups(REQUEST *request, void *instance, UNUSED void *thread, void *ctx)
 {
-	rlm_rcode_t		rcode = RLM_MODULE_NOOP;
-	VALUE_PAIR		*check_tmp = NULL, *reply_tmp = NULL, *sql_group = NULL;
-	rlm_sql_grouplist_t	*head = NULL, *entry = NULL;
+	rlm_sql_t const *inst = instance;
+	rlm_sql_thread_group_t *group_ctx = ctx;
 
-	char			*expanded = NULL;
-	int			rows;
+	rad_assert(ctx);
 
 	rad_assert(request->packet != NULL);
 
+	/*
+	 * Because of the new async model, this function gets called multiple times
+	 * at different stages of the group processing.
+	 * Simply jump to where processing needs to resume.
+	 */
+	switch (group_ctx->next_step) {
```

Comment:
The problem with that is that it makes the code impossible to read.
That's what I started doing, but it was impossible to follow the overall logic.
It's mainly because we need to exit some functions from within a loop and therefore resume from that same point.